src/main/java/com/vmware/vipclient/i18n/messages/service/PatternService.java

```diff
@@ -30,63 +33,90 @@ public JSONObject getPatternsByCategory(String locale, String category) {
-                // and
-                logger.info("Cache pattern!\n\n");
-                new PatternCacheService().addPatterns(key, patterns);
-            }
+        if (patterns != null) {
+            logger.debug("Find pattern from cache for language [{}], region [{}]!", language, region);
+            return patterns;
+        }
+        patterns = getPatternsFromBundle(language, region);
+        if (patterns != null) {
+            logger.debug("Find the pattern for language [{}], region [{}].\n", language, region);// [datetime]
+            // and
+            new PatternCacheService().addPatterns(key, patterns);
+            logger.debug("Pattern is cached for language [{}], region [{}]!\n\n", language, region);
+            return patterns;
+        }
+        if (!LocaleUtility.isDefaultLocale(new Locale(language, region))) {
+            logger.info("Can't find pattern for language [{}] region [{}], look for English pattern as fallback!", language, region);
+            patterns = getPatterns(ConstantsKeys.EN);
         }
         return patterns;
     }
 
     private JSONObject getPatternsFromBundle(String locale) {
         JSONObject patterns = null;
         if (LocaleUtility.isDefaultLocale(locale)) {
-            logger.debug("Got pattern from local bundle!");
+            logger.debug("Look for pattern from local bundle for locale [{}]!", locale);
             patterns = new LocalPatternOpt()
-                    .getPatternsByLocale(ConstantsKeys.EN);
+                    .getEnPatterns(ConstantsKeys.EN);
         } else {
-            if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-            	patterns = new RemotePatternOpt().getPatternsByLocale(locale);
+            if (VIPCfg.getInstance().getMsgOriginsQueue().get(0) == DataSourceEnum.VIP) {
+                logger.debug("Look for pattern from Singleton Service for locale [{}]!", locale);
+                patterns = new RemotePatternOpt().getPatternsByLocale(locale);
             } else {
-                patterns = new LocalPatternOpt().getPatternsByLocale(locale);
+                logger.debug("Look for pattern from local bundle for locale [{}]!", locale);
+                String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
+                logger.debug("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
+                patterns = new LocalPatternOpt().getPatternsByLocale(normalizedLocale);
             }
         }
         return patterns;
     }
 
     private JSONObject getPatternsFromBundle(String language, String region) {
         JSONObject patterns = null;
-        if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-        	patterns = new RemotePatternOpt().getPatternsByLocale(language, region);
+        Locale locale = new Locale(language, region);
+        if (LocaleUtility.isDefaultLocale(locale)) {
```

Comment:
(Same comment as above)
This does not follow the client workflow design. Always try to fetch from the service first even for the default locale, so remove the if-statement block.